modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java

```diff
@@ -571,22 +608,31 @@ public KBArticle expireKBArticle(
 			long userId, long resourcePrimKey, ServiceContext serviceContext)
 		throws PortalException {
 
-		KBArticle kbArticle = getLatestKBArticle(
-			resourcePrimKey, WorkflowConstants.STATUS_ANY);
+		if (!hasKBArticleLock(userId, resourcePrimKey)) {
+			lockKBArticle(userId, resourcePrimKey);
+		}
 
-		if (kbArticle.isDraft() || kbArticle.isPending() ||
-			kbArticle.isScheduled()) {
+		try {
+			KBArticle kbArticle = getLatestKBArticle(
+				resourcePrimKey, WorkflowConstants.STATUS_ANY);
 
-			return kbArticle;
-		}
+			if (kbArticle.isDraft() || kbArticle.isPending() ||
+				kbArticle.isScheduled()) {
+
+				return kbArticle;
+			}
 
-		kbArticle.setExpirationDate(new Date());
+			kbArticle.setExpirationDate(new Date());
 
-		kbArticleLocalService.updateKBArticle(kbArticle);
+			kbArticleLocalService.updateKBArticle(kbArticle);
 
-		return updateStatus(
-			userId, resourcePrimKey, WorkflowConstants.STATUS_EXPIRED,
-			serviceContext);
+			return updateStatus(
+				userId, resourcePrimKey, WorkflowConstants.STATUS_EXPIRED,
+				serviceContext);
+		}
+		finally {
+			unlockKBArticle(resourcePrimKey);
+		}
```

Comment:
This change would imply calling the Unlock method in Saving MVC action command.

Reply:
Why not? 

If you check the DL logic it works this way:

1: When you start the method we check if the current user has a lock:

`hasKBArticleLock(PrincipalThreadLocal.getUserId(), resourcePrimKey)`

If not, a new lock is been created (if possible, since lock logic controls the use cases to create it or not - expired, same user..):

`lockKBArticle(PrincipalThreadLocal.getUserId(), resourcePrimKey);`

So, when you finish the method, we know that this lock belongs to this user and it could be removed safely.

Do yo see any inconvenience in this workflow?

Reply:
The use case not handled correctly by the current implementation is the following:

* There is a higher level logic calling the KB Article API. This logic wants to lock a KB article to make some modifications on it
* When starting a lock is acquired on the KB Article
* During the task several calls to the `updateKBArticle` are done
* When the task is finished the KB Article is unlocked.

I think this a reasonable way of thinking about the locking mechanism. However, with the proposed implementation the lock is erroneously released after the first update.

Reply:
1. During the task several calls to the updateKBArticle are done

Reply:
Hi @adolfopa can you please give your opinion on this? Is the use case I describe something to take into account? Or can we just safely ignore it? Thanks

Reply:
For consistency with DM, the behavior from the edit screen should be (roughly):
1. When clicking on cancel, unlock.
2. When publishing/scheduling, unlock.
3. When saving as draft, keep lock active.

For this we need two "update" operations in the KB API: one that doesn't remove the lock, and another one (analogous to DM's updateAndCheckIn()) that updates and removes the lock. This will be called from the UI/action.